luigi/scheduler.py

```diff
@@ -567,7 +567,8 @@ def prune(self):
 
         for task in self._state.get_active_tasks():
             self._state.fail_dead_worker_task(task, self._config, assistant_ids)
-            if task.id not in necessary_tasks and self._state.prune(task, self._config):
+            removed = self._state.prune(task, self._config)
+            if removed and task.id not in necessary_tasks:
```

Comment:
Exactly @Tarrasch. I figured it would be safer (and more clear) to break out the method call in case someone comes along and adds another condition that would prevent the method from being called, i.e.

```python
if some_other_false_condition and self._state.prune(task, self._config) and task.id not in necessary_tasks:
   # ...
```